compiler/rustc_middle/src/query/keys.rs

```diff
@@ -26,9 +26,11 @@ pub trait Key: Sized {
     //      r-a issue: <https://github.com/rust-lang/rust-analyzer/issues/13693>
     type CacheSelector;
 
+    type LocalKey;
+
     /// Given an instance of this key, what crate is it referring to?
     /// This is used to find the provider.
-    fn query_crate_is_local(&self) -> bool;
+    fn as_local_key(&self) -> Option<Self::LocalKey>;
```

Comment:
This is the essence of the change -- instead of `query_crate_is_local` returning a `bool`, we map a key to its local form and use that if it succeeds, calling the local provider.

Reply:
Alternatively, we could split this out into a separate trait that only needs to be implemented on keys of `separate_provide_extern` queries... most keys don't ever need `as_local_key` called on it, anyways.